src/omv/ports/stm32/omv_spi.c

```diff
@@ -152,18 +152,80 @@ static void omv_spi_callback(SPI_HandleTypeDef *hspi) {
         spi = omv_spi_descr_all[5];
     #endif
     }
-    if (spi == NULL) {
-        return;
-    }
+    return spi;
+}
 
+static void omv_spi_callback_error_check(SPI_HandleTypeDef *hspi, omv_spi_t *spi) {
     if (hspi->ErrorCode != HAL_SPI_ERROR_NONE) {
         spi->xfer_flags |= OMV_SPI_XFER_FAILED;
         spi->xfer_error = hspi->ErrorCode;
         omv_spi_transfer_abort(spi);
+    } else {
+        spi->xfer_flags |= OMV_SPI_XFER_COMPLETE;
+    }
+}
+
+static void *omv_spi_callback_buf(omv_spi_t *spi) {
+    uint8_t *buf = (uint8_t *) spi->descr->pRxBuffPtr;
+
+    if (buf == NULL) {
+        buf = (uint8_t *) spi->descr->pTxBuffPtr;
+    }
+
+    if (spi->xfer_flags & OMV_SPI_XFER_HALF) {
+        return buf;
+    } else {
+        uint32_t size = spi->descr->RxXferSize;
+
+        if (!size) {
+            size = spi->descr->TxXferSize;
+        }
+
+        return buf + ((size * ((spi->xfer_flags & OMV_SPI_XFER_16_BIT) ? 2 : 1)) / 2);
+    }
+}
+
+static void omv_spi_callback(SPI_HandleTypeDef *hspi) {
+    omv_spi_t *spi = omv_spi_get_descr(hspi);
+    if (spi == NULL) {
+        return;
     }
 
+    omv_spi_callback_error_check(hspi, spi);
+    spi->xfer_flags &= ~OMV_SPI_XFER_HALF;
+
     if (spi->callback) {
-        spi->callback(spi, spi->userdata);
+        spi->callback(spi, spi->userdata, omv_spi_callback_buf(spi));
+    }
+}
+
+static void omv_spi_half_callback(SPI_HandleTypeDef *hspi) {
```

Comment:
I don't think we need the half callback, or those extra functions, this should work:
```suggestion
static void omv_spi_callback(SPI_HandleTypeDef *hspi) {
    omv_spi_t *spi = omv_spi_get_descr(hspi);
    if (spi == NULL) {
        return;
    }

    if (hspi->ErrorCode != HAL_SPI_ERROR_NONE) {
        spi->xfer_flags |= OMV_SPI_XFER_FAILED;
        spi->xfer_error = hspi->ErrorCode;
        omv_spi_transfer_abort(spi);
    } else {
        spi->xfer_flags |= OMV_SPI_XFER_COMPLETE;
    }

    if (spi->callback) {
        uint8_t *buf = (spi->descr->pRxBuffPtr) ? spi->descr->pRxBuffPtr : spi->descr->pTxBuffPtr;
        if (config->dma_flags & OMV_SPI_DMA_DOUBLE) {
            if (spi->xfer_flags & OMV_SPI_XFER_HALF) {
                uint32_t size = (spi->descr->RxXferSize) ? (spi->descr->RxXferSize) : spi->descr->TxXferSize;
                buf += ((size * ((spi->xfer_flags & OMV_SPI_XFER_16_BIT) ? 2 : 1)) / 2);
            }
            spi->xfer_flags ^= OMV_SPI_XFER_HALF;
        }
        spi->callback(spi, spi->userdata, buf);
    }
}
```

Reply:
Done